src/rulesets/Base/array.jl

```diff
@@ -1,3 +1,32 @@
+#####
+##### constructors
+#####
+
+ChainRules.@non_differentiable (::Type{T} where {T<:Array})(::UndefInitializer, args...)
+
+function rrule(::Type{T}, x::AbstractArray) where {T<:Array}
+    project_x = ProjectTo(x)
+    Array_pullback(ȳ) = (NoTangent(), project_x(ȳ))
+    return T(x), Array_pullback
+end
+
+#####
+##### `vect`
+#####
+
+@non_differentiable Base.vect()
+
+# Don't worry about projection here. The data passes straight through, so if a cotangent has
+# the wrong type for some reason, it must be the fault of another rule somewhere.
+function rrule(::typeof(Base.vect), X::T...) where {T}
+    l = length(X)
+    function vect_pullback(ȳ)
+        X̄ = ntuple(n -> ȳ[n], l)
+        return (NoTangent(), X̄...)
```

Comment:
I think this will infer a bit better:
```suggestion
function rrule(::typeof(Base.vect), x::Vararg{T,N}) where {T,N}
    function vect_pullback(ȳ)
        return (NoTangent(), NTuple{N}(ȳ)...)
```
Trying the existing one at the REPL, it forgets that the length is a constant:
```
julia> @code_warntype rrule(Base.vect, 1,2,3)[2]([1,2,3])
MethodInstance for (::var"#vect_pullback#17"{Int64})(::Vector{Int64})
  from (::var"#vect_pullback#17")(ȳ) in Main at REPL[14]:3
Arguments
  #self#::var"#vect_pullback#17"{Int64}
  ȳ::Vector{Int64}
Locals
  #16::var"#16#18"{Vector{Int64}}
  X̄::Tuple{Vararg{Int64}}
Body::Tuple{NoTangent, Vararg{Int64}}
```

Reply:
This is more elegant than the equivalent thing I went for -- will adopt this.